pkg/networkservice/common/heal/client.go

```diff
@@ -62,95 +61,86 @@ type healClient struct {
-		eventReceiver:     nil, // This is intentionally nil
-		recvEventExecutor: serialize.NewExecutor(),
-		chainContext:      ctx,
+		ctx:            ctx,
+		client:         client,
+		onHeal:         onHeal,
+		heals:          make(map[string]func()),
+		connections:    make(map[string]*networkservice.Connection),
+		updateExecutor: serialize.NewExecutor(),
 	}
 
 	if rv.onHeal == nil {
 		rv.onHeal = addressof.NetworkServiceClient(rv)
 	}
 
-	rv.init()
-
 	return rv
 }
 
-func (f *healClient) init() {
-	if f.eventReceiver != nil {
+func (f *healClient) monitorLoop() {
+	go f.eventLoopOnce.Do(func() {
+		for {
+			select {
+			case <-f.ctx.Done():
+				return
+			default:
+			}
+			recv, err := f.client.MonitorConnections(f.ctx, &networkservice.MonitorScopeSelector{}, grpc.WaitForReady(true))
+			if err != nil {
+				continue
+			}
+			f.eventloop(recv)
+		}
+	})
+}
+
+// Should only be called inside monitorLoop
+func (f *healClient) eventloop(recv networkservice.MonitorConnection_MonitorConnectionsClient) {
+	for {
 		select {
-		case <-f.eventReceiver.Context().Done():
+		case <-f.ctx.Done():
 			return
 		default:
-			f.eventReceiver = nil
 		}
-	}
-
-	logrus.Info("Creating new eventReceiver")
-
-	recv, err := f.client.MonitorConnections(f.chainContext, &networkservice.MonitorScopeSelector{}, grpc.WaitForReady(true))
-	if err != nil {
-		f.updateExecutor.AsyncExec(func() {
-			f.init()
-		})
-		return
-	}
-	f.eventReceiver = recv
-	f.recvEventExecutor.AsyncExec(f.recvEvent)
-}
-
-func (f *healClient) recvEvent() {
-	select {
-	case <-f.eventReceiver.Context().Done():
-		return
-	default:
-		event, err := f.eventReceiver.Recv()
-		f.updateExecutor.AsyncExec(func() {
-			if err != nil {
-				for id, r := range f.requestors {
-					r()
-					delete(f.reported, id)
+		event, err := recv.Recv()
+		if err != nil {
+			f.updateExecutor.AsyncExec(func() {
+				for id, heal := range f.heals {
+					heal()
+					delete(f.connections, id)
 				}
-				f.init()
-				return
-			}
-
+			})
+			break
+		}
+		f.updateExecutor.AsyncExec(func() {
 			switch event.GetType() {
 			case networkservice.ConnectionEventType_INITIAL_STATE_TRANSFER:
-				f.reported = event.GetConnections()
+				f.connections = event.GetConnections()
 				if event.GetConnections() == nil {
-					f.reported = map[string]*networkservice.Connection{}
+					f.connections = map[string]*networkservice.Connection{}
 				}
 
 			case networkservice.ConnectionEventType_UPDATE:
 				for _, conn := range event.GetConnections() {
-					f.reported[conn.GetId()] = conn
+					f.connections[conn.GetId()] = conn
 				}
 
 			case networkservice.ConnectionEventType_DELETE:
 				for _, conn := range event.GetConnections() {
-					delete(f.reported, conn.GetId())
+					delete(f.connections, conn.GetId())
 				}
 			}
 
-			for id, request := range f.requestors {
-				if _, ok := f.reported[id]; !ok {
-					request()
+			for id, heal := range f.heals {
+				if _, ok := f.connections[id]; !ok {
+					heal()
```

Comment:
@edwarnicke I do not think it is a good idea to do a heal on every update for every connection, 
Assume just one is added we receive update, and we do a bunch of requests for any of others.

I think we need to heal only on delete.

Update's case fo monorepo will propogate event to previous client, and in case of context changes( DNS change for example), every client was able to immediately react, but now we will need to wait, so for Updates it is also doesn't make seance to call Heal.

Initial transfer also not require to call heal, if and only if we do not have connections on received list.

Thoughts?

Reply:
I agree that healing every connection on every update would not be a good idea.  This doesn't do that.

f.connections reflects what has been reported to us by our peer.
f.heals reflects what we believe we have in terms of connections.

When we receive an update, act on f.connections to have it reflect that update.

At the end, we go through the connections we think we have (f.heals) and 'heal' the ones that have not been reported to us in f.connections (note the !ok in the if statement)

If our peer reports an INITIAL_STATE transfer and its missing a connection we feel we should have, we heal that.
If our peer DELETEs a connection, we heal that.

Does that make sense?

Reply:
Ok, sounds good.